### PR TITLE
New Actors Quickstart

### PR DESCRIPTION
**Pull request by paulyuk**
# Description

New quickstart for Actors building block.  Details are in the proposal.  

## Issue reference

#775 

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

* [x] The quickstart code compiles correctly
* [x] You've tested new builds of the quickstart if you changed quickstart code
* [x] You've updated the quickstart's README if necessary
* [ ] If you have changed the steps for a quickstart be sure that you have updated the automated validation accordingly. All of our quickstarts have annotations that allow them to be executed automatically as code. For more information see [mechanical-markdown](https://github.com/dapr/mechanical-markdown#readme). For user guide with examples see [Examples](https://github.com/dapr/mechanical-markdown/tree/main/examples#mechanical-markdown-by-example). 
    - Tracked with issue https://github.com/dapr/quickstarts/issues/836

### CODE REVIEW

File: actors/csharp/sdk/README.md [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+```bash
+cd actors/csharp/sdk/service
+dotnet build
+```
+
+Run the `SmartDevice.Service`, which will start service itself and the Dapr sidecar:
+
+```bash
+dapr run --app-id actorservice --app-port 5001 --dapr-http-port 3500 --components-path ../../../resources -- dotnet run --urls=http://localhost:5001/
```
Comment (msfussell) [inline actors/csharp/sdk/README.md:39]:
```suggestion
dapr run --app-id actorservice --app-port 5001 --dapr-http-port 3500 --resources-path ../../../resources -- dotnet run --urls=http://localhost:5001/
```

Comment (msfussell) [reply actors/csharp/sdk/README.md:39]: @paulyuk @marcduiker - One minor change whilst testing. I would suggest two things
1) Worth adding a register reminder timer on the controller that is called back after say 30s to turn off the alarms. This shows how to use a reminder
2) Work with @hhunter-ms to draw a diagram of the controller and the actors docs PR https://github.com/dapr/docs/pull/3309/files

Comment (paulyuk) [reply actors/csharp/sdk/README.md:39]: great suggestions @msfussell adding now. thanks for review

**Review by paulyuk — COMMENTED** [review]
All feedback addressed.  Docs handled separately in another issue (https://github.com/dapr/docs/pull/3309/files). 

@msfussell ready to merge

**Comment by paulyuk** [comment]
@hhunter-ms  fyi

File: actors/csharp/sdk/README.md [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+When you ran the client app:
+
+1. Two `SmartDetectorActor` actors are created and initialized with Id, Location, and Status="Ready"
+2. The `DetectSmokeAsync` method of `SmartDetectorActor` 1 is called.
+3. The `TriggerAlarmForAllDetectors` method of `ControllerActor` is called.
+4. The `SoundAlarm` methods of `SmartDetectorActor` 1 and 2 are called.
+5. The `ControllerActor` also creates a reminder to `ClearAlarm` after 15 seconds using `RegisterReminderAsync`
+
+
+Looking at the code, `SmartDetectorActor` objects are created in the client application and initialized with object state with `ActorProxy.Create<ISmartDevice>(actorId, actorType)` and then `proxySmartDevice.SetDataAsync(data)`.  These objects are re-entrant and will hold on to the state as shown by `proxySmartDevice.GetDataAsync()`.
```
Comment (msfussell) [inline actors/csharp/sdk/README.md:103]:
```suggestion
Looking at the code, `SmartDetectorActor` objects are created in the client application and initialized with object state with `ActorProxy.Create<ISmartDevice>(actorId, actorType)` and then `proxySmartDevice.SetDataAsync(data)`.  These objects are re-entrant and hold the state as shown by `proxySmartDevice.GetDataAsync()`.
```

File: actors/csharp/sdk/service/SmokeDetectorActor.cs [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+        return "Success";
+    }
+
+    /// <summary>
+    /// Get MyData from actor's private state store
+    /// </summary>
+    /// <return>the user-defined MyData which is stored into state store as "my_data" state</return>
+    public async Task<SmartDeviceData> GetDataAsync()
+    {
+        // Gets state from the state store.
```
Comment (msfussell) [inline actors/csharp/sdk/service/SmokeDetectorActor.cs:66]:
```suggestion
        // Gets current state from the state store.
```

File: actors/csharp/sdk/service/SmokeDetectorActor.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+        return Task.CompletedTask;
+    }
+
+    /// <summary>
+    /// Set MyData into actor's private state store
+    /// </summary>
+    /// <param name="data">the user-defined MyData which will be stored into state store as "device_data" state</param>
+    public async Task<string> SetDataAsync(SmartDeviceData data)
+    {
+        // Data is saved to configured state store implicitly after each method execution by Actor's runtime.
```
Comment (msfussell) [inline actors/csharp/sdk/service/SmokeDetectorActor.cs:50]:
```suggestion
        // Data is saved to configured state store *implicitly* after each method execution by Actor's runtime.
```

File: actors/csharp/sdk/service/SmokeDetectorActor.cs [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        // Data can also be saved explicitly by calling this.StateManager.SaveStateAsync();
```
Comment (msfussell) [inline actors/csharp/sdk/service/SmokeDetectorActor.cs:51]:
```suggestion
        // Data can also be saved *explicitly* by calling this.StateManager.SaveStateAsync();
```

File: actors/csharp/sdk/README.md [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+
+### What happened
+
+When you ran the client app:
+
+1. Two `SmartDetectorActor` actors are created and initialized with Id, Location, and Status="Ready"
+2. The `DetectSmokeAsync` method of `SmartDetectorActor` 1 is called.
+3. The `TriggerAlarmForAllDetectors` method of `ControllerActor` is called.
+4. The `SoundAlarm` methods of `SmartDetectorActor` 1 and 2 are called.
+5. The `ControllerActor` also creates a reminder to `ClearAlarm` after 15 seconds using `RegisterReminderAsync`
```
Comment (msfussell) [inline actors/csharp/sdk/README.md:100]:
```suggestion
5. The `ControllerActor` also creates a durable reminder to call `ClearAlarm` after 15 seconds using `RegisterReminderAsync`
```

File: actors/csharp/sdk/service/ControllerActor.cs [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            var actorId = new ActorId(deviceId);
+            var proxySmartDevice = ProxyFactory.CreateActorProxy<ISmartDevice>(actorId, "SmokeDetectorActor");
+            await proxySmartDevice.SoundAlarm();
+        }
+
+        // Register a reminder to refresh and clear alarm state every 15 seconds
+        await this.RegisterReminderAsync("AlarmRefreshReminder", null, TimeSpan.FromSeconds(15), TimeSpan.FromSeconds(15));
+    }
+
+    public async Task ReceiveReminderAsync(string reminderName, byte[] state, TimeSpan dueTime, TimeSpan period)
```
Comment (msfussell) [inline actors/csharp/sdk/service/ControllerActor.cs:68]:
```suggestion
    // Callback method for all Reminders. Check the Reminder name for which timer fired.
    public async Task ReceiveReminderAsync(string reminderName, byte[] state, TimeSpan dueTime, TimeSpan period)
```

**Review by msfussell — COMMENTED** [review]
Looks good. Here are a few minors updates.

**Comment by paulyuk** [comment]
@msfussell thanks for tweaks, resolved

**Pull request merged**